### ISSUE TITLE
Narrowing mapped types to exclude undefined values

### ISSUE THREAD

**Issue by jg** [Microsoft/TypeScript#13528]
# Problem

**TypeScript Version:**  2.1.4

**Code**

(--strictNullChecks)

```ts
interface Store {
  value: string | undefined
}

export function get<K extends keyof Store>(
  store: Store,
  fieldName: K
): Store[K] {
  if (store[fieldName] == undefined) throw(fieldName + " is undefined!");

  return store[fieldName]
}

const store: Store = { value: "stuff" }

const result = get(store, "value").replace("stuff", "otherstuff")
```

**Expected behavior:**

(It's possible to narrow the resulting type of `get` to exclude undefined values)

**Actual behavior:**

Type error is reported: get(a, "value") is possibly undefined.

# Proposal

I'd propose an operator which could be used to express that the mapped type does not contain `undefined`. Consider the use of bang (!) operator below:

```ts
export function get<K extends keyof Store>(
  store: Store,
  fieldName: K
): Store[K]! {
  if (store[fieldName] == undefined) throw(fieldName + " is undefined!");

  return store[fieldName]
}
```

**Comment by aluanhaddad** [comment]
I believe this is what you are looking for #12578

**Comment by jg** [comment]
Yep
